packages/fuel-indexer-schema/src/db/queries.rs

```diff
@@ -94,19 +96,105 @@ impl UserQuery {
-                    query = [query, joins.join(" ")].join(" ");
-                }
+    /// Returns a SQL query that contains the requested results and a PageInfo object.
+    fn create_query_with_pageinfo(
+        &self,
+        db_type: &DbType,
+        selections_str: String,
+        joins_str: String,
+        limit: u64,
+    ) -> String {
+        // In order to create information about pagination, we need to calculate
+        // values according to the amount of records, current offset, and requested
+        // limit. To avoid sending additional queries for every request sent to
+        // the API, we leverage a common table expression (CTE) which is a table
+        // that exists only for the duration of the query and allows us to refer
+        // to its result set.
+        match db_type {
+            db_type @ DbType::Postgres => {
+                let json_selections_str =
+                    self.get_json_selections_from_cte(db_type).join(",");
+
+                let selection_cte = format!(
+                    r#"WITH selection_cte AS (
+                        SELECT json_build_object({}) AS {}
+                        FROM {}.{}
+                        {}
+                        {}
+                        {}),"#,
+                    selections_str,
+                    self.entity_name,
+                    self.namespace_identifier,
+                    self.entity_name,
+                    joins_str,
+                    self.query_params.get_filtering_expression(db_type),
+                    self.query_params.get_ordering_modififer(db_type),
+                );
 
-                query = [query, self.query_params.to_sql(db_type)].join(" ");
+                let total_count_cte =
+                    "total_count_cte AS (SELECT COUNT(*) as count FROM selection_cte)"
+                        .to_string();
+
+                let offset = self.query_params.offset.unwrap_or(0);
+                let alias = self.alias.clone().unwrap_or(self.entity_name.clone());
+
+                let selection_query = format!(
+                    r#"SELECT json_build_object(
+                        'page_info', json_build_object(
+                            'has_next_page', (({limit} + {offset}) < (SELECT count from total_count_cte)),
+                            'limit', {limit},
+                            'offset', {offset},
+                            'pages', ceil((SELECT count from total_count_cte)::float / {limit}::float),
+                            'total_count', (SELECT count from total_count_cte)
+                        ),
+                        '{alias}', (
+                            SELECT json_agg(item)
+                            FROM (
+                                SELECT {json_selections_str} FROM selection_cte
+                                LIMIT {limit} OFFSET {offset}
+                            ) item
+                        )
+                    );"#
+                );
```

Comment:
I am sure you have tested this case, but what happens the alias value is the `self.entity_name`, and then is the entity name is used in the placeholder `{'{alias}'}`, does it effect the query? Maybe it makes no difference, but perhaps we should remove this value altogether from being used, if no `alias` has been set.

Reply:
Nope, it doesn't affect the results in any way. It just sets the list of objects as a value to the key of the entity name:
```
query {
  block(order: { desc: height}, first: 5, offset: 5) {
    id
    hash
    height
    timestamp
  }
}
```
```
{
  "data": {
    "block": [
        ...
    ],
    "pageInfo": {
        ...
    }
  }
}
```

I think it should stay as any downstream users will be able to refer to the objects themselves or the `PageInfo` object with just a key. Without it, both parts would have to be elements in a list which would require users to do additional object checking on their end.